src/download/pre_msg_metadata.rs

```diff
@@ -0,0 +1,248 @@
+use anyhow::{Context as _, Result};
+use num_traits::ToPrimitive;
+use serde::{Deserialize, Serialize};
+
+use crate::context::Context;
+use crate::log::warn;
+use crate::message::Message;
+use crate::message::Viewtype;
+use crate::param::{Param, Params};
+
+/// Metadata contained in Pre-Message that describes the Post-Message.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct PreMsgMetadata {
+    /// size of the attachment in bytes
+    pub(crate) size: u64,
+    /// Real viewtype of message
+    pub(crate) viewtype: Viewtype,
+    /// the original file name
+    pub(crate) filename: String,
+    /// Dimensions: width and height of image or video
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) dimensions: Option<(i32, i32)>,
```

Comment:
Can be renamed to `wh` for more consistency with `Param::{Width,Height}` and to state that the width goes first